### PR TITLE
Added support to choose configMap name

### LINKED ISSUE #1634: Enrichers to configmap
Its not possible to change the name of configmap in tag `configMap`, always is `xmlconfig`.

```
$ oc get configmap
NAME        DATA      AGE
xmlconfig   1         20m
```
### Description
Include in project the config below and run `mvn fabric8:deploy`
```
<configuration>
                            <resources>
                                <configMap>
                                    <entries>
                                        <entry>
                                            <file>src/main/resources/application.properties</file>
                                        </entry>
                                    </entries>
                                </configMap>
                            </resources>
                        </configuration>
```

### Info

* f-m-p version : 4.1.0
* Maven version (`mvn -v`) :
```
Apache Maven 3.5.2
Maven home: /usr/share/maven
Java version: 1.8.0_171, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-openjdk-amd64/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.15.0-29deepin-generic", arch: "amd64", family: "unix"
```
* Kubernetes / OpenShift setup and version (`oc version`) : 
```
oc v3.10.45
kubernetes v1.10.0+b81c8f8
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://192.168.42.210:8443
kubernetes v1.11.0+d4cacc0
```

* Minishift version (`minishift version`) : 
```
minishift v1.33.0+ba29431
```

Looking the classes [1] [2] I propose create a property `name` in class [2]. I will try create a local version and if works I open a pull request.

[1] - https://github.com/fabric8io/fabric8-maven-plugin/blob/master/enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/ConfigMapEnricher.java
[2] - https://github.com/fabric8io/fabric8-maven-plugin/blob/master/core/src/main/java/io/fabric8/maven/core/config/ConfigMap.java

### PR DESCRIPTION
**Pull request by h3nrique**
Fixes #1634

### CODE REVIEW

File: core/src/main/java/io/fabric8/maven/core/config/ConfigMap.java [hunk 1]
```diff
@@ -15,11 +15,15 @@
...
 package io.fabric8.maven.core.config;
 
+import org.apache.maven.plugins.annotations.Parameter;
+
 import java.util.ArrayList;
 import java.util.List;
 
 public class ConfigMap {
 
+    @Parameter
```
Comment (rohanKanojia) [inline core/src/main/java/io/fabric8/maven/core/config/ConfigMap.java]: I don't think that is required.

**Comment by rohanKanojia** [comment]
strange, looking at code I think it should work. I would pull your branch and test on my machine.

**Comment by rohanKanojia** [comment]
Oops, a test is failing. Maybe that can give us an idea if we're missing something.

**Comment by h3nrique** [comment]
Strange.. I will check here..

Thanks

---
Atenciosamente,
*Paulo Henrique Alves*
*It is not necessary to do extraordinary things to get extraordinary
results. (W. Buffett)*


Em qua, 15 de mai de 2019 às 15:21, Rohan Kumar <notifications@github.com>
escreveu:

**Comment by h3nrique** [comment]
The error is in
method ConfigMapEnricher.getConfigMapFromXmlConfiguration(), this method is
not loading the configs for ConfigMap doesnot exists.

I've made a change that will not throws a NPE, the tests will pass now. :)

[image: DeepinScreenshot_20190515154859.png]

---
Atenciosamente,
*Paulo Henrique Alves*
*It is not necessary to do extraordinary things to get extraordinary
results. (W. Buffett)*


Em qua, 15 de mai de 2019 às 15:30, Paulo Henrique Alves <
paulohenriqueas13@gmail.com> escreveu:

**Comment by rohanKanojia** [comment]
@h3nrique : Works for me for this configuration:

```
                        <configMap>
                          <name>rohan-configmap</name>
                          <entries>
                            <entry>
                              <name>A</name>
                              <value>B</value>
                            </entry>
                           </entries>
                        </configMap>
```

What configuration were you trying for which it wasn't working? Could you please share? 
Screenshot from 2019-05-17 21-15-15

**Comment by h3nrique** [comment]
@rohanKanojia I've identified that when uses variable (ex: ${project.artifactId}) the configmap It's not created.

I'm working in this here.

**Comment by rohanKanojia** [comment]
@lordofthejars : @paoloantinori  is trying to create configmap using this configuration:

```
<configuration>
    <resources>
        <configMap>
            <name>${project.artifactId}</name>
            <entries>
                <entry>
                    <file>src/main/resources/application-openshift.properties</file>
                </entry>
            </entries>
        </configMap>
    </resources>
</configuration>
```

but it doesn't seem to be getting picked up. Could you please take a look into this if you have time? Is this configuration supported?

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/ConfigMapEnricher.java [hunk 2]
```diff
@@ -91,9 +92,12 @@ private void addConfigMapFromXmlConfigurations(KubernetesListBuilder builder) {
         final Map<String, String> configMapFromConfiguration;
         try {
             configMapFromConfiguration = createConfigMapFromConfiguration(configMap);
-            if(!configMapFromConfiguration.isEmpty() && !checkIfItemExists(builder, "xmlconfig")) {
+            String configMapName = (configMap == null || configMap.getName() == null || configMap.getName().trim().isEmpty()) ? "xmlconfig" : configMap.getName().trim();
+            log.debug("configMapName :: ".concat(configMapName));
+
+            if(!configMapFromConfiguration.isEmpty() && !checkIfItemExists(builder, configMapName)) {
```
Comment (rohanKanojia) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/ConfigMapEnricher.java]: I think the reason why config map is not getting created is due to `checkIfItemExists(builder, configMapName)` returning true always(ConfigMap name was hardcoded earlier, that's why it returned false, and at this point already created Deployment & Service are also created with same name). I think we should remove this method or maybe change it check for kind also

Comment (rohanKanojia) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/ConfigMapEnricher.java]: Or maybe change checkIfItemExists to this:
```
    private boolean checkIfItemExists(KubernetesListBuilder builder, String name) {
        return builder.buildItems().stream().anyMatch(item -> item.getKind().equals("CongifMap") && item.getMetadata().getName().equals(name));
    }
```

**Comment by rohanKanojia** [comment]
@h3nrique : Is it working now for you? for your `application.properties` use case?

**Comment by h3nrique** [comment]
Now It's working as well, I made a update that you say.
Thanks.

Em qua, 22 de mai de 2019 11:52, Rohan Kumar <notifications@github.com>
escreveu:

**Comment by rohanKanojia** [comment]
@h3nrique : cool, can you please mark this draft ready for merge? I think we can merge it then :+1:

**Comment by rohanKanojia** [comment]
@h3nrique : Sorry, I forgot to tell you. Could you please add a line to CHANGELOG.md regarding this?

**Comment by sleipnir** [comment]
Great this is exactly what I need

**Pull request merged**